### DISCUSSION TITLE
Cannot Set Thunk Functions

### DISCUSSION THREAD

**Discussion by j3s53** [NationalSecurityAgency/ghidra#4049] [Q&A] [answered]
Ghidra fails to set any new thunked functions to external libraries.

I'm working on an arm program with several thunk functions that point to external libraries. I have several libraries in my symbol tree and they all have the correct imports associated with them in the thunk functions. However, after removing and re-adding a library in my imports and redefining the thunk function, every time I "Set Thunked Function..." to the added external library in the form of libname::funcname, it throws the error `Invalid thunk reference address or name specified: libname::funcname`. How do I get Ghidra to recognize the imported library again?

![image](https://user-images.githubusercontent.com/100794036/156418223-8732bfe9-bec4-4e98-8aa2-f1693046d33c.png)

![image](https://user-images.githubusercontent.com/100794036/156418262-cb8d1ed4-611e-44ef-bae0-867306241892.png)

Side note: I also am not able to get my working program to update the imported function signatures if I change a function in the imported library.

**Comment by ghidra1** [comment, marked as answer]
Does the specified library `libwhatever` appear as a Library under the Imports tab of the Symbol Tree (it should have a gold cube icon)?  If so, do the Function `funcname` appear as a child node of this library in the Symbol Tree?  The Function node in the Symbol Tree must have the `f` icon (a yellow dot icon indicate a non-function location).  If these conditions are not met it will not find the external symbol with the symbol path you have specified.  How did you go about recreating the Library and its associated Functions?

**Reply by j3s53** [reply]
When I first opened the binary to analyze, all external functions were listed under the flat `<EXTERNAL>` library in the Symbol Tree Imports folder. I then created a library `libwhatever` and `Set External Program` pointing to the library in my Ghidra project. Everything worked as you described after running the `FixupELFExternalSymbolsScript.java` script.

However, I removed and re-added the `libwhatever` library in the Symbol Tree `Imports` folder (and made sure to `Set External Program`). Now `funcname` is permanently gone. I tried to reanalyze the current program to generate the missing symbols but they did not reappear, not even under `<EXTERNAL>`.

**Reply by ghidra1** [reply]
I assume it is too late to Undo.  When you removed the library all of its child symbols (i.e., external locations/functions) also get removed as will as any references to them.  While you can manually re-add all of these under the newly created library, it may be easier to re-import to a new Program file, analyze and re-run the script.  Once this is done you can use the Program Diff to help migrate listing details from old to new Program file.  You can't specify a thunked function until the referenced external function exists (i.e., has been re-added under the specified library).  External symbols are initially created by the ELF Importer and not analysis.

**Reply by j3s53** [reply]
I managed to correctly create the association with `Create External Location` from the library context menu, thank you! 

Is there a way to push function signatures (and/or changes) from the external library to the Symbol Tree and thus the Thunk signatures? When I set the thunk fuction, it did not bring over the function signature.
